main/src/ca/uwaterloo/flix/language/phase/EarlyTreeShaker.scala

```diff
@@ -0,0 +1,438 @@
+
+    case Expression.InvokeMethod(_, exp, args, _, _, _, _) =>
+      visitExp(exp) ++ visitExps(args)
+
+    case Expression.InvokeStaticMethod(_, args, _, _, _, _) =>
+      visitExps(args)
+
+    case Expression.GetField(_, exp, _, _, _, _) =>
+      visitExp(exp)
+
+    case Expression.PutField(_, exp1, exp2, _, _, _, _) =>
+      visitExp(exp1) ++ visitExp(exp2)
+
+    case Expression.GetStaticField(_, _, _, _, _) =>
+      Set.empty
+
+    case Expression.PutStaticField(_, exp, _, _, _, _) =>
+      visitExp(exp)
+
+    case Expression.NewObject(_, _, _, _, _) =>
+      Set.empty
+
+    case Expression.NewChannel(exp, _, _, _, _) =>
+      visitExp(exp)
+
+    case Expression.GetChannel(exp, _, _, _, _) =>
+      visitExp(exp)
+
+    case Expression.PutChannel(exp1, exp2, _, _, _, _) =>
+      visitExp(exp1) ++ visitExp(exp2)
+
+    case Expression.SelectChannel(rules, default, _, _, _, _) =>
+      visitExps(rules.map(_.chan)) ++ visitExps(rules.map(_.exp)) ++ default.map(visitExp).getOrElse(Set.empty)
+
+    case Expression.Spawn(exp, _, _, _, _) =>
+      visitExp(exp)
+
+    case Expression.Lazy(exp, _, _) =>
+      visitExp(exp)
+
+    case Expression.Force(exp, _, _, _, _) =>
+      visitExp(exp)
+
+    case Expression.FixpointConstraintSet(cs, _, _, _) =>
+      visitConstraints(cs)
+
+    case Expression.FixpointLambda(_, exp, _, _, _, _, _) =>
+      visitExp(exp)
+
+    case Expression.FixpointMerge(exp1, exp2, _, _, _, _, _) =>
+      visitExp(exp1) ++ visitExp(exp2)
+
+    case Expression.FixpointSolve(exp, _, _, _, _, _) =>
+      visitExp(exp)
+
+    case Expression.FixpointFilter(_, exp, _, _, _, _) =>
+      visitExp(exp)
+
+    case Expression.FixpointInject(exp, _, _, _, _, _) =>
+      visitExp(exp)
+
+    case Expression.FixpointProject(_, exp, _, _, _, _) =>
+      visitExp(exp)
+
+    case Expression.Reify(_, _, _, _, _) =>
+      Set.empty
+
+    case Expression.ReifyType(_, _, _, _, _, _) =>
+      Set.empty
+
+    case Expression.ReifyEff(_, exp1, exp2, exp3, _, _, _, _) =>
+      visitExp(exp1) ++ visitExp(exp2) ++ visitExp(exp3)
+
+    case Expression.Do(sym, exps, _, _, _) =>
+      visitExps(exps) // Add OpSym to reachable syms?
+
+    case Expression.Resume(exp, _, _) =>
+      visitExp(exp)
+
+    case Expression.TryWith(exp, _, rules, _, _, _, _) =>
+      visitExp(exp) ++ visitExps(rules.map(_.exp))
+
+    case Expression.Without(exp, _, _, _, _, _) =>
+      visitExp(exp)
+
+  }
+
+  /**
+    * Returns the function symbols reachable from `es`.
+    */
+  private def visitExps(es: List[Expression]): Set[ReachableSym] = es.map(visitExp).fold(Set())(_ ++ _)
+
+
+  /**
+    * Represents a super type for reachable symbols in the AST.
+    */
+  sealed trait ReachableSym
+
+  object ReachableSym {
+
+    case class DefnSym(defnSym: Symbol.DefnSym) extends ReachableSym
+
+    case class SigSym(sigSym: Symbol.SigSym) extends ReachableSym
+
+    case class ClassSym(classSym: Symbol.ClassSym) extends ReachableSym
```

Comment:
I'm not sure I even need this case class. I think the relevant classes (and defs) can be derived from SigSyms and DefnSyms. Agreed?